code/week8/main.cpp

```diff
@@ -0,0 +1,111 @@
+}
+
+void printIntVec(std::vector<int> vec) {
+    std::cout << "printing vector " << std::endl;
+    for(std::vector<int>::iterator it = vec.begin();it != vec.end(); it++) {
+        std::cout << *it << ", ";
+    }
+    std::cout << "\n\n";
+}
+
+void printInt(int i) {
+    std::cout << "looking at " << i << std::endl;
+}
+
+class Parent {
+public:
+    virtual void do_something() {std::cout << "BASE" << std::endl;}
+};
+
+class Child : public Parent {
+public:
+    virtual void do_something() {std::cout << "CHILD" << std::endl;}
+};
+
+void callDoSomething(Parent p) {
+    p.do_something();
+}
+
+int main() {
+    using namespace std;
+
+    /******** FUNCTION POINTERS ********/
+
+    // functions are just lines of code stored somewhere in memory
+    // that means that you can get the address of a function just like anything else using the & operator
+    // RETURN_TYPE (*NAME) (PARAM_TYPES);
+    int (*func_ptr) (int) = &addTwo;
+    int (*name) (int, int) = &add;
+
+    // although a function can be implicitly converted into a pointer by the compiler
+    // so the & is not required
+    int (*func_ptr2) (int) = addThree;
+
+    // to use add ()
+    cout << "value of addTwo(2) = " << func_ptr(2) << endl;
+
+    /**************** IGNORE IF YOU DO NOT UNDERSTAND ****************/
+    // since a function pointer is also a pointer you can dereference it but this is not necessary
+    // in fact when we dereference the pointer we get the function
+    // which as mentioned above can be implicitly converted into a pointer by the compiler
+    // this means that we can add an arbitrary number of * and this code still works
+
+    cout << "calling addTwo(2) = " << (*************** func_ptr)(2) << endl;
+
+    /**************** EXAMPLE CODE RESUMED ***************************/
+
+    /******** LAMBDAS ********/
+    cout << "\n\n";
+
+    vector<int> example_vec = {1,2,3,4,5};
+    // for_each does whatever function is given to all values will not change the values
+    // http://www.cplusplus.com/reference/algorithm/for_each/?kw=for_each
+    for_each(example_vec.begin(), example_vec.end(), printInt);
+
+    cout << "\n\n";
+
+    // using lambda
+    // [CAPTURE_VAR] (TYPE NAME, ...) -> RETURN_TYPE {CODE};
+    for_each(example_vec.begin(), example_vec.end(), [](int num) -> void {std::cout << "printing from lambda " << num << std::endl;});
+
+    // transform is like for each but it replaces the value with what is returned
+    // http://www.cplusplus.com/reference/algorithm/transform/?kw=transform
+    transform(example_vec.begin(), example_vec.end(), example_vec.begin(), [](int num) -> int {return num + 2;});
+    printIntVec(example_vec);
+
+    // capture allows us to pass in another var as a constant to the lambda
+    // this makes the variable effectively constant while in the method
+    int constant = 7;
+    transform(example_vec.begin(), example_vec.end(), example_vec.begin(), [constant](int num) -> int {return num + constant;});
```

Comment:
Not sure if we want to cover this, but I think it would be nice to include a closure example as well, since that's the standard ways variable capture work in other languages. 

In cpp, I think it's done by `[&](stuff)`

So maybe 

```
int i = 0;
([&]() -> void {i++;})();
cout << i << endl;
```
Did not try compiling that but that's the idea at least.